cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java

```diff
@@ -0,0 +1,140 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data.tools;
+
+import co.cask.cdap.api.dataset.DatasetAdmin;
+import co.cask.cdap.api.dataset.DatasetSpecification;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.datafabric.DefaultDatasetNamespace;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.lib.hbase.AbstractHBaseDataSetAdmin;
+import co.cask.cdap.data2.dataset2.lib.table.hbase.HBaseTableAdmin;
+import co.cask.cdap.data2.transaction.queue.QueueAdmin;
+import co.cask.cdap.data2.util.hbase.HBaseTableUtil;
+import co.cask.cdap.data2.util.hbase.TableId;
+import co.cask.cdap.proto.Id;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HTableDescriptor;
+import org.apache.hadoop.hbase.client.HBaseAdmin;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Handles upgrade for System and User Datasets
+ */
+public class DatasetUpgrader extends AbstractUpgrader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MDSUpgrader.class);
+
+  private final CConfiguration cConf;
+  private final Configuration hConf;
+  private final LocationFactory locationFactory;
+  private final QueueAdmin queueAdmin;
+  private final HBaseTableUtil hBaseTableUtil;
+  private final DatasetFramework namespacedFramework;
+
+  @Inject
+  public DatasetUpgrader(CConfiguration cConf, Configuration hConf, LocationFactory locationFactory,
+                         QueueAdmin queueAdmin, HBaseTableUtil hBaseTableUtil,
+                         @Named("namespacedDSFramework") DatasetFramework namespacedFramework) {
+
+    super(locationFactory);
+    this.cConf = cConf;
+    this.hConf = hConf;
+    this.locationFactory = locationFactory;
+    this.queueAdmin = queueAdmin;
+    this.hBaseTableUtil = hBaseTableUtil;
+    this.namespacedFramework = namespacedFramework;
+  }
+
+  @Override
+  public void upgrade() throws Exception {
+    // Upgrade system dataset
+    upgradeSystemDatasets(namespacedFramework);
+
+    // Upgrade all user hbase tables
+    upgradeUserTables();
+
+    // Upgrade all queue and stream tables.
+    queueAdmin.upgrade();
+  }
+
+  private void upgradeSystemDatasets(DatasetFramework framework) throws Exception {
+
+    // Upgrade all datasets in system namespace
+    Id.Namespace systemNamespace = Id.Namespace.from(Constants.SYSTEM_NAMESPACE);
+    for (DatasetSpecification spec : framework.getInstances(systemNamespace)) {
```

Comment:
I will have a look but the history shows that this line was modified by you recently. Did you checked when you modified it ?